File: backend/app/core/permissions.py
import enum

from app.models.enums import Role


class Permission(str, enum.Enum):
    OVERVIEW_VIEW = "overview.view"
    PERFORMANCE_VIEW = "performance.view"
    OPERATIONAL_INTELLIGENCE_VIEW = "operational_intelligence.view"
    OPERATIONAL_IMPACT_CONTRIBUTE = "operational_impact.contribute"
    OUTPUTS_VIEW = "outputs.view"
    REPORTS_CREATE = "reports.create"
    REPORTS_DOWNLOAD = "reports.download"


# Rol, permission'ların sabit bir paketidir — inheritance zinciri yok, her rol kendi tam
# permission kümesini açıkça listeler. Yeni bir rol eklemek (ör. ileride READ_ONLY/REPORT_USER)
# sadece bu dict'e bir satır eklemekle olur; kod içinde `role == X` veya rol seviyesi
# karşılaştırması YAPILMAZ — her yetki kararı `permission in permissions` ile verilir.
ROLE_PERMISSIONS: dict[Role, frozenset[Permission]] = {
    Role.FOREMAN: frozenset(
        {
            Permission.OVERVIEW_VIEW,
            Permission.PERFORMANCE_VIEW,
        }
    ),
    Role.SUPERVISOR: frozenset(
        {
            Permission.OVERVIEW_VIEW,
            Permission.PERFORMANCE_VIEW,
            Permission.OPERATIONAL_INTELLIGENCE_VIEW,
            Permission.OPERATIONAL_IMPACT_CONTRIBUTE,
        }
    ),
    Role.OPERATIONS_MANAGER: frozenset(
        {
            Permission.OVERVIEW_VIEW,
            Permission.PERFORMANCE_VIEW,
            Permission.OPERATIONAL_INTELLIGENCE_VIEW,
            Permission.OPERATIONAL_IMPACT_CONTRIBUTE,
            Permission.OUTPUTS_VIEW,
            Permission.REPORTS_CREATE,
            Permission.REPORTS_DOWNLOAD,
        }
    ),
}


ROLE_LABELS_TR: dict[Role, str] = {
    Role.FOREMAN: "Formen",
    Role.SUPERVISOR: "Şef",
    Role.OPERATIONS_MANAGER: "Operasyon Yöneticisi",
}


def permissions_for_role(role: Role) -> frozenset[Permission]:
    return ROLE_PERMISSIONS.get(role, frozenset())

File: frontend/src/auth/permissions.ts
// Backend'deki tek doğruluk kaynağıyla (app/core/permissions.py) birebir eşleşen sabitler.
// Permission -> rol eşlemesi burada TUTULMAZ: rol ve izinler her zaman /auth/me'den gelir,
// frontend kendi başına hesaplamaz (bkz. PermissionContext).
export type Permission =
  | "overview.view"
  | "performance.view"
  | "operational_intelligence.view"
  | "operational_impact.contribute"
  | "outputs.view"
  | "reports.create"
  | "reports.download";

export type Role = "FOREMAN" | "SUPERVISOR" | "OPERATIONS_MANAGER";

export const ROLE_LABELS_TR: Record<Role, string> = {
  FOREMAN: "Formen",
  SUPERVISOR: "Şef",
  OPERATIONS_MANAGER: "Operasyon Yöneticisi",
};
